src/main/java/com/mongodb/spark/sql/connector/schema/BsonDocumentToRowConverter.java

```diff
@@ -0,0 +1,501 @@
+  }
+
+  private float convertToFloat(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return (float) convertToBsonNumber(fieldName, dataType, bsonValue).doubleValue();
+  }
+
+  private double convertToDouble(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return convertToBsonNumber(fieldName, dataType, bsonValue).doubleValue();
+  }
+
+  private int convertToInteger(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return convertToBsonNumber(fieldName, dataType, bsonValue).intValue();
+  }
+
+  private short convertToShort(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return (short) convertToBsonNumber(fieldName, dataType, bsonValue).intValue();
+  }
+
+  private long convertToLong(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return convertToBsonNumber(fieldName, dataType, bsonValue).longValue();
+  }
+
+  private BigDecimal convertToDecimal(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    BsonNumber bsonNumber = convertToBsonNumber(fieldName, dataType, bsonValue);
+    switch (bsonNumber.getBsonType()) {
+      case DOUBLE:
+        return BigDecimal.valueOf(bsonNumber.doubleValue());
+      case INT32:
+      case INT64:
+        return BigDecimal.valueOf(bsonNumber.longValue());
+      default:
+        return bsonNumber.decimal128Value().bigDecimalValue();
+    }
+  }
+
+  private byte convertToByte(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    return (byte) convertToBsonNumber(fieldName, dataType, bsonValue).longValue();
+  }
+
+  private String convertToString(final BsonValue bsonValue) {
+    switch (bsonValue.getBsonType()) {
+      case STRING:
+        return bsonValue.asString().getValue();
+      case SYMBOL:
+        return bsonValue.asSymbol().getSymbol();
+      case DOCUMENT:
+        return bsonValue.asDocument().toJson(jsonWriterSettings);
+      default:
+        String value = new BsonDocument("v", bsonValue).toJson(jsonWriterSettings);
+        // Strip down to just the value
+        value = value.substring(6, value.length() - 1);
+        // Remove unnecessary quotes of BsonValues converted to Strings.
+        // Such as BsonBinary base64 string representations
+        if (value.startsWith("\"") && value.endsWith("\"")) {
+          value = value.substring(1, value.length() - 1);
+        }
+        return value;
+    }
+  }
+
+  private BsonNumber convertToBsonNumber(
+      final String fieldName, final DataType dataType, final BsonValue bsonValue) {
+    switch (bsonValue.getBsonType()) {
+      case DATE_TIME:
+        return new BsonInt64(bsonValue.asDateTime().getValue());
+      case TIMESTAMP:
+        return new BsonInt64(bsonValue.asTimestamp().getTime() * 1000L); // normalize to millis
```

Comment:
Consider using `SECONDS.toMillis(bsonValue.asTimestamp().getTime())` as it is more expressive and does not require a clarification comment.

Reply:
:+1: